packages/data/src/types.ts

```diff
@@ -1,40 +1,66 @@
-export type WPDataFunctionOrGeneratorArray = Array< Function | Generator >;
-export type WPDataFunctionArray = Array< Function >;
+type MapOf< T > = { [ name: string ]: T };
 
-export interface WPDataAttachedStore {
-	getSelectors: () => WPDataFunctionArray;
-	getActions: () => WPDataFunctionArray;
+export type ActionCreator = Function | Generator;
+export type Resolver = Function | Generator;
+export type Selector = Function;
+
+export type AnyConfig = StoreConfig< any, any >;
+
+export interface StoreInstance< Config extends AnyConfig > {
+	getSelectors: () => SelectorsOf< Config >;
```

Comment:
I think using the `SelectorsOf` type here is incorrect. `getSelectors` returns bound selectors, without the `state` argument, while the `Selectors` in `Config` are unbound.

Similarly, `getActions` returns bound actions, and here it affects only the return value. For example, a bound action will never return a generator/iterator.

As long as the typedef is just `Function` this is all fine, but as soon as you try to specify the types more precisely there will be trouble.

Generally, for a particular store, `getSelectors` and `getActions` return generic interfaces that don't use any Redux or `@wordpress/data` types. For example:
```ts
interface InterfaceSelectors {
  getSingleEnableItem( itemType: string, scope: string ): string | null;
  isMultipleEnabledItemEnabled( itemType: string, scope: string, item: string ): boolean | undefined;
}

interface InterfaceActions {
  setSingleEnableItem( itemType: string, scope: string ): void;
  setMultipleEnableItem( itemType: string, scope: string, item: string, isEnable: boolean ): void;
}
```
are both interfaces that any JS object could implement, they are just divided into "readers" and "writers" parts.

The `StoreConfig` types should be defined in terms of these public APIs, not the other way around. `SelectorsOf<InterfaceSelectors, TState>` should prepend a `state` argument to every method of `InterfaceSelectors`, and `ActionCreatorsOf<InterfaceActions>` should produce action creator types that correspond to the signatures of the bound action methods.

Reply:
thank you @jsnajdr - "getSelectors returns bound selectors, without the state argument, while the Selectors in Config are unbound" on this I agree but I don't want to address this yet. currently the config doesn't change the state of how these wind through the types because of what you identified - `Function`

in fact I'm trying to do this piecemeal to break the problem down and limit the size of any individual change.

the purpose therefore of `SelectorsOf<Config>` is to set us up so that later we can fill out that function without changing the config type. we can curry or bind the types of those selectors that are inferred through type mapping and `SelectorsOf<Config>` remains correct, but much more specific

> SelectorsOf<InterfaceSelectors, TState> should prepend a state argument to every method of InterfaceSelectors, and ActionCreatorsOf<InterfaceActions> should produce action creator types that correspond to the signatures of the bound action methods.

Exactly. I have working code for these, but in order to get there I've had to make more dramatic changes plus I've been dealing with issues where quirky things are happening and I think it's because TypeScript has hit its heuristic limit and given up.

> The StoreConfig types should be defined in terms of these public APIs, not the other way around.

What I'm trying hard to do is define the store config implicitly and entirely through inference. So far it's working quite well and, for example, in order to fully-type a package like `edit-post` all we need to do is rename its files to `.ts` and add a `tsconfig.json`.

---

If you think that the `SelectorsOf<Config>` type is too misleading I could rewrite it to only keep the names

```ts
type SelectorsOf<Config extends AnyConfig> = Config extends StoreConfig<any, infer Selectors>
	? { [ name in key Selectors ]: Function }
	: never;
```

that seems better while not changing things too much. if you like this I'll commit it.

Reply:
relaxed the constraints to the above in 0d8e9b4830